Oh no, build fails on FreeBSD:

```
./Modules/socketmodule.c:2116:28: error: incompatible pointer to integer conversion initializing 'unsigned short' with an expression of type 'char[32]' [-Wint-conversion]
 2116 |             unsigned short dev = _BT_HCI_MEMB(addr, dev);
      |                            ^     ~~~~~~~~~~~~~~~~~~~~~~~
./Modules/socketmodule.c:2122:37: error: array type 'char[32]' is not assignable
 2122 |             _BT_HCI_MEMB(addr, dev) = dev;
      |             ~~~~~~~~~~~~~~~~~~~~~~~ ^
```

---

Unrelated failures.

ARM64 Windows PR and ARM64 Windows Non-Debug PR: pythoninfo steps crash with exit code `-1073741819`.

s390x Fedora Stable Clang PR, s390x RHEL8 PR, s390x RHEL9 PR: test_binary_header() of test_tools.test_msgfmt.CompilationTest failed. I don't know why, but it's unrelated to this change.

--- a/Modules/socketmodule.c
+++ b/Modules/socketmodule.c
@@ -2107,6 +2107,21 @@ getsockaddrarg(PySocketSockObject *s, PyObject *args,
             straddr = PyBytes_AS_STRING(args);
             if (setbdaddr(straddr, &_BT_HCI_MEMB(addr, bdaddr)) < 0)
                 return 0;
+#elif defined(__FreeBSD__)
+            _BT_HCI_MEMB(addr, family) = AF_BLUETOOTH;
+            if (!PyBytes_Check(args)) {
+                PyErr_Format(PyExc_OSError, "%s: "
+                             "wrong node format", caller);
+                return 0;
+            }
+            const char *straddr = PyBytes_AS_STRING(args);
+            size_t len = PyBytes_GET_SIZE(args);
+            if (len >= sizeof(_BT_HCI_MEMB(addr, node))) {
+                PyErr_Format(PyExc_OSError, "%s: "
+                             "node too long", caller);
+                return 0;
+            }
+            strcpy(_BT_HCI_MEMB(addr, node), straddr);
 #else  /* __NetBSD__ || __DragonFly__ */
             _BT_HCI_MEMB(addr, family) = AF_BLUETOOTH;
             unsigned short dev = _BT_HCI_MEMB(addr, dev);

--- a/Modules/socketmodule.c
+++ b/Modules/socketmodule.c
@@ -1521,11 +1521,15 @@ makesockaddr(SOCKET_T sockfd, struct sockaddr *addr, size_t addrlen, int proto)
             struct sockaddr_hci *a = (struct sockaddr_hci *) addr;
 #if defined(__NetBSD__) || defined(__DragonFly__)
             return makebdaddr(&_BT_HCI_MEMB(a, bdaddr));
-#else /* __NetBSD__ || __DragonFly__ */
+#elif defined(__FreeBSD__)
+            char *node = _BT_HCI_MEMB(a, node);
+            size_t len = strnlen(node, sizeof(_BT_HCI_MEMB(a, node)));
+            return PyBytes_FromStringAndSize(node, (Py_ssize_t)len);
+#else
             PyObject *ret = NULL;
             ret = Py_BuildValue("i", _BT_HCI_MEMB(a, dev));
             return ret;
-#endif /* !(__NetBSD__ || __DragonFly__) */
+#endif
         }
 
 #if !defined(__FreeBSD__)
@@ -2116,12 +2120,18 @@ getsockaddrarg(PySocketSockObject *s, PyObject *args,
             }
             const char *straddr = PyBytes_AS_STRING(args);
             size_t len = PyBytes_GET_SIZE(args);
-            if (len >= sizeof(_BT_HCI_MEMB(addr, node))) {
+            if (strlen(straddr) != len) {
+                PyErr_Format(PyExc_OSError, "%s: "
+                             "node contains embedded null character", caller);
+                return 0;
+            }
+            if (len > sizeof(_BT_HCI_MEMB(addr, node))) {
                 PyErr_Format(PyExc_OSError, "%s: "
                              "node too long", caller);
                 return 0;
             }
-            strcpy(_BT_HCI_MEMB(addr, node), straddr);
+            strncpy(_BT_HCI_MEMB(addr, node), straddr,
+                    sizeof(_BT_HCI_MEMB(addr, node)));
 #else  /* __NetBSD__ || __DragonFly__ */
             _BT_HCI_MEMB(addr, family) = AF_BLUETOOTH;
             unsigned short dev = _BT_HCI_MEMB(addr, dev);

--- a/Modules/socketmodule.c
+++ b/Modules/socketmodule.c
@@ -2121,12 +2121,12 @@ getsockaddrarg(PySocketSockObject *s, PyObject *args,
             const char *straddr = PyBytes_AS_STRING(args);
             size_t len = PyBytes_GET_SIZE(args);
             if (strlen(straddr) != len) {
-                PyErr_Format(PyExc_OSError, "%s: "
+                PyErr_Format(PyExc_ValueError, "%s: "
                              "node contains embedded null character", caller);
                 return 0;
             }
             if (len > sizeof(_BT_HCI_MEMB(addr, node))) {
-                PyErr_Format(PyExc_OSError, "%s: "
+                PyErr_Format(PyExc_ValueError, "%s: "
                              "node too long", caller);
                 return 0;
             }

